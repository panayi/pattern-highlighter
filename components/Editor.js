import React from "react";
import SimpleCodeEditor from 'react-simple-code-editor'
import Highlight, { defaultProps } from 'prism-react-renderer'
import theme from 'prism-react-renderer/themes/vsLight'

const editorStyles = {
  height: '100%',
  fontFamily: '"Dank Mono", "Fira Code", monospace',
  ...theme.plain
}

const CharacterRenderer = (props) => {
  const { char, index, offset, styles } = props;

  const activeStyles = React.useMemo(() => {
    const globalIndex = offset + props.index;

    return styles.filter(({ startIndex, endIndex }) => (
      globalIndex >= startIndex && globalIndex < endIndex
    ))
  }, [index, offset, styles])

  return activeStyles.reduce((acc, item) => (
    <span style={item.style}>
      {acc}
    </span>
  ), [char])
}

const TokenRenderer = (props) => {
  const { styles, offset, children, ...rest } = props;

  return (
    <span {...rest}>
      {children.split('').map((char, index) => (
        <CharacterRenderer key={index} index={index} offset={offset} char={char} styles={styles} />
      ))}
    </span>
  )
}

const HighlightCode = ({ code, styles }) => {
  let offset = 0;

  return (
    <Highlight {...defaultProps} theme={theme} code={code} language="json">
      {({ tokens, getLineProps, getTokenProps }) => (
        <>
          {tokens.map((line, i) => {
            const lineOutput = (
              <div {...getLineProps({ line, key: i })}>
                {line.map((token, key) => {
                  const tokenProps = getTokenProps({ token, key });

                  const tokenOutput = (
                    <TokenRenderer {...tokenProps} offset={offset} styles={styles} />
                  );
                  
                  if (!token.empty) {
                    offset += tokenProps.children.length;
                  }

                  return tokenOutput;
                })}
              </div>
            );
            
            // new-line character
            offset += 1;

            return lineOutput;
          })}
        </>
      )}
    </Highlight>
  );
}

const noop = () => null;

const Editor = ({ content = '', styles = [], onChange = noop }) => {
  return (
    <SimpleCodeEditor
      value={content}
      onValueChange={onChange}
      highlight={code => <HighlightCode code={code} styles={styles} />}
      padding={10}
      style={editorStyles}
    />
  )
};

export default Editor;