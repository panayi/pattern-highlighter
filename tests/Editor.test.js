import '@testing-library/jest-dom';
import { render, fireEvent } from '@testing-library/react';
import Editor from '../components/Editor';

describe('[component] Editor', () => {
  it('should render without any passed props', () => {
    const component = render(<Editor />);
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should render with all passed props', () => {
    const onChange = () => 1;
    const styles = [
      {
        startIndex: 1,
        endIndex: 3,
        style: {
          backgroundColor: 'red'
        }
      }
    ]
    const component = render(<Editor content="foobar" styles={styles} onChange={onChange} />);
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should render code', () => {
    const { container } = render(<Editor content="const foo = 1;" />);
    const pre = container.querySelector('pre');
    expect(pre.textContent).toEqual('const foo = 1;');
  });

  it('should set textarea value', () => {
    const { container } = render(<Editor content="const foo = 1;" />);
    const textarea = container.querySelector('textarea');
    expect(textarea.value).toEqual('const foo = 1;');
  });

  it('should call onChange on value change', () => {
    const onChange = jest.fn();
    const { container } = render(<Editor content="const foo = 1;" onChange={onChange} />);
    const textarea = container.querySelector('textarea');
    fireEvent.change(textarea, { target: { value: 'const foo = 1; const bar = 2;' } });
    expect(onChange.mock.calls.length).toBe(1);
    expect(onChange.mock.calls[0][0]).toBe('const foo = 1; const bar = 2;');
  });

  it('should apply styles', () => {
    const styles = [
      {
        startIndex: 1,
        endIndex: 3,
        style: {
          backgroundColor: 'red'
        }
      },
      {
        startIndex: 6,
        endIndex: 8,
        style: {
          backgroundColor: 'green'
        }
      }
    ]
    const content = `012\n\n567`
    const { container } = render(<Editor content={content} styles={styles} />);
    const spans = container.querySelectorAll('pre span:not(.token)');

    expect(spans.length).toEqual(4);
    expect(spans[0].style.backgroundColor).toEqual('red');
    expect(spans[0].textContent).toEqual('1');
    expect(spans[1].style.backgroundColor).toEqual('red');
    expect(spans[1].textContent).toEqual('2');
    expect(spans[2].style.backgroundColor).toEqual('green');
    expect(spans[2].textContent).toEqual('6');
    expect(spans[3].style.backgroundColor).toEqual('green');
    expect(spans[3].textContent).toEqual('7');
  });

  it('should apply multiple styles', () => {
    const styles = [
      {
        startIndex: 1,
        endIndex: 4,
        style: {
          backgroundColor: 'red'
        }
      },
      {
        startIndex: 2,
        endIndex: 3,
        style: {
          backgroundColor: 'green'
        }
      }
    ]
    const content = `lorem`
    const { container } = render(<Editor content={content} styles={styles} />);
    const spans = container.querySelectorAll('pre span:not(.token)')
    const nestedSpans = container.querySelectorAll('pre span:not(.token) > span:not(.token)');

    expect(spans.length).toEqual(4);
    expect(nestedSpans.length).toEqual(1);
    expect(nestedSpans[0].style.backgroundColor).toEqual('red');
    expect(nestedSpans[0].parentElement.style.backgroundColor).toEqual('green');
    expect(nestedSpans[0].textContent).toEqual('r');
  });
});