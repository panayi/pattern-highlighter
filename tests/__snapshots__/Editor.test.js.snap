// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[component] Editor should render with all passed props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; text-align: left; box-sizing: border-box; padding: 0px; overflow: hidden; height: 100%; font-family: \\"Dank Mono\\", \\"Fira Code\\", monospace; color: rgb(214, 222, 235); background-color: rgb(1, 22, 39);"
  >
    <textarea
      autocapitalize="off"
      autocomplete="off"
      autocorrect="off"
      class="npm__react-simple-code-editor__textarea"
      data-gramm="false"
      spellcheck="false"
      style="margin: 0px; border: 0px; background: none; box-sizing: inherit; display: inherit; font-family: inherit; font-style: inherit; font-variant-ligatures: inherit; font-weight: inherit; letter-spacing: inherit; line-height: inherit; tab-size: inherit; text-indent: inherit; text-rendering: inherit; text-transform: inherit; white-space: pre-wrap; word-break: keep-all; overflow-wrap: break-word; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; resize: none; overflow: hidden; padding: 10px 10px 10px 10px;"
    >
      foobar
    </textarea>
    <pre
      aria-hidden="true"
      style="margin: 0px; border: 0px; background: none; box-sizing: inherit; display: inherit; font-family: inherit; font-style: inherit; font-variant-ligatures: inherit; font-weight: inherit; letter-spacing: inherit; line-height: inherit; tab-size: inherit; text-indent: inherit; text-rendering: inherit; text-transform: inherit; white-space: pre-wrap; word-break: keep-all; overflow-wrap: break-word; position: relative; pointer-events: none; padding: 10px 10px 10px 10px;"
    >
      <div
        class="token-line"
        style="color: rgb(214, 222, 235);"
      >
        <span
          class="token plain"
        >
          f
          <span
            style="background-color: red;"
          >
            o
          </span>
          <span
            style="background-color: red;"
          >
            o
          </span>
          bar
        </span>
      </div>
    </pre>
    <style
      type="text/css"
    >
      
/**
 * Reset the text fill color so that placeholder is visible
 */
.npm__react-simple-code-editor__textarea:empty {
  -webkit-text-fill-color: inherit !important;
}

/**
 * Hack to apply on some CSS on IE10 and IE11
 */
@media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
  /**
    * IE doesn't support '-webkit-text-fill-color'
    * So we use 'color: transparent' to make the text transparent on IE
    * Unlike other browsers, it doesn't affect caret color in IE
    */
  .npm__react-simple-code-editor__textarea {
    color: transparent !important;
  }

  .npm__react-simple-code-editor__textarea::selection {
    background-color: #accef7 !important;
    color: transparent !important;
  }
}

    </style>
  </div>
</DocumentFragment>
`;

exports[`[component] Editor should render without any passed props 1`] = `
<DocumentFragment>
  <div
    style="position: relative; text-align: left; box-sizing: border-box; padding: 0px; overflow: hidden; height: 100%; font-family: \\"Dank Mono\\", \\"Fira Code\\", monospace; color: rgb(214, 222, 235); background-color: rgb(1, 22, 39);"
  >
    <textarea
      autocapitalize="off"
      autocomplete="off"
      autocorrect="off"
      class="npm__react-simple-code-editor__textarea"
      data-gramm="false"
      spellcheck="false"
      style="margin: 0px; border: 0px; background: none; box-sizing: inherit; display: inherit; font-family: inherit; font-style: inherit; font-variant-ligatures: inherit; font-weight: inherit; letter-spacing: inherit; line-height: inherit; tab-size: inherit; text-indent: inherit; text-rendering: inherit; text-transform: inherit; white-space: pre-wrap; word-break: keep-all; overflow-wrap: break-word; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; resize: none; overflow: hidden; padding: 10px 10px 10px 10px;"
    />
    <pre
      aria-hidden="true"
      style="margin: 0px; border: 0px; background: none; box-sizing: inherit; display: inherit; font-family: inherit; font-style: inherit; font-variant-ligatures: inherit; font-weight: inherit; letter-spacing: inherit; line-height: inherit; tab-size: inherit; text-indent: inherit; text-rendering: inherit; text-transform: inherit; white-space: pre-wrap; word-break: keep-all; overflow-wrap: break-word; position: relative; pointer-events: none; padding: 10px 10px 10px 10px;"
    >
      <div
        class="token-line"
        style="color: rgb(214, 222, 235);"
      >
        <span
          class="token plain"
          style="display: inline-block;"
        >
          

        </span>
      </div>
    </pre>
    <style
      type="text/css"
    >
      
/**
 * Reset the text fill color so that placeholder is visible
 */
.npm__react-simple-code-editor__textarea:empty {
  -webkit-text-fill-color: inherit !important;
}

/**
 * Hack to apply on some CSS on IE10 and IE11
 */
@media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
  /**
    * IE doesn't support '-webkit-text-fill-color'
    * So we use 'color: transparent' to make the text transparent on IE
    * Unlike other browsers, it doesn't affect caret color in IE
    */
  .npm__react-simple-code-editor__textarea {
    color: transparent !important;
  }

  .npm__react-simple-code-editor__textarea::selection {
    background-color: #accef7 !important;
    color: transparent !important;
  }
}

    </style>
  </div>
</DocumentFragment>
`;
